runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -441,21 +448,23 @@ resolveRefToObject(J9VMThread *currentThread, J9ConstantPool *ramConstantPool, U
 }
 
 J9Method *
-lookupMethod(J9VMThread *currentThread, J9Class *resolvedClass, J9JNINameAndSignature *nas, J9Class *callerClass, UDATA lookupOptions)
+lookupMethod(J9VMThread *currentThread, J9Class *resolvedClass, J9NameAndSignature *nas, J9Class *callerClass, UDATA lookupOptions)
 {
 	J9Method *result = NULL;
 
 	/* If looking for a MethodHandle polymorphic INL method, allow any caller signature. */
 	if (resolvedClass == J9VMJAVALANGINVOKEMETHODHANDLE(currentThread->javaVM)) {
-		if ((0 == strcmp(nas->name, "linkToVirtual"))
-		|| (0 == strcmp(nas->name, "linkToStatic"))
-		|| (0 == strcmp(nas->name, "linkToSpecial"))
-		|| (0 == strcmp(nas->name, "linkToInterface"))
-		|| (0 == strcmp(nas->name, "invokeBasic"))
+		J9UTF8 *nameUTF = nas->name;
+		U_8 *nameData = J9UTF8_DATA(nameUTF);
+		U_16 nameLength = J9UTF8_LENGTH(nameUTF);
+
+		if ((J9UTF8_LITERAL_EQUALS(nameData, nameLength, "linkToVirtual"))
+		|| (J9UTF8_LITERAL_EQUALS(nameData, nameLength, "linkToStatic"))
+		|| (J9UTF8_LITERAL_EQUALS(nameData, nameLength, "linkToSpecial"))
+		|| (J9UTF8_LITERAL_EQUALS(nameData, nameLength, "linkToInterface"))
+		|| (J9UTF8_LITERAL_EQUALS(nameData, nameLength, "invokeBasic"))
 		) {
-			nas->signature = NULL;
-			nas->signatureLength = 0;
-
+			J9UTF8_SET_LENGTH(nas->signature, 0);
```

Comment:
This destructively modifies the input parameter.  Is that specified in the method's contract?  Will callers expect their signature to be 0 length after a call to `lookupMethod`?

Reply:
the `MH.linkTo*` methods have polymorphic signatures, so `lookupMethod` should only check class & name, in this case, the `J9_LOOK_PARTIAL_SIGNATURE` flag is set to skip signature check

Reply:
The concern isn't with `lookupMethod`'s use of the signature, but with the callers of `lookupMethod`.  Do they all take into account that **their** data may be modified in this way?  It's a non-obvious modification of the passed in data and it would be good to define that as the contract or to not modify the callers data.

Reply:
I've refectored the code so that the caller will directly pass in name/sig as J9UTF8 so the the `lookupMethod` call will be constructing the J9NAS and set J9_LOOK_DIRECT_NAS flag. Also added nullSignature to replace the signature passed by caller without modifying it.